### ISSUE TITLE
Massive slowdown when using flycheck or racket-xp-mode

### ISSUE THREAD

**Issue by hjones2199** [greghendershott/racket-mode#473]
When navigating or editing with flycheck-mode or racket-xp-mode enabled i'm having serious performance problems that are only noticable when im actively working in the racket buffer. I have tried racket-mode-start-faster and it helped with repl startup times but has not effected the .rkt file slowdowns.

I noticed with racket-xp-mode on and flycheck-mode off, the slowdowns seem to be significantly worse when the cursor is passing through a variable/function name compared to being at the beginning of a line or on a literal.

<details>
```
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
                 (20200513 1519)
                 "Racket editing, REPL, and more"
                 ((emacs
                   (25 1))
                  (faceup
                   (0 0 2))
                  (pos-tip
                   (20191127 1028)))
                 nil nil "/home/username/.emacs.d/elpa/racket-mode-20200513.1519"
                 ((:url . "https://www.racket-mode.com/")
                  (:maintainer "Greg Hendershott")
                  (:authors
                   ("Greg Hendershott")))
                 nil))
((emacs-version "26.1")
 (system-type gnu/linux)
 (major-mode racket-mode)
 (racket--el-source-dir "/home/username/.emacs.d/elpa/racket-mode-20200513.1519/")
 (racket--rkt-source-dir "/home/username/.emacs.d/elpa/racket-mode-20200513.1519/racket/")
 (racket-program "racket")
 (racket-command-timeout 10)
 (racket-xp-after-change-refresh-delay 1)
 (racket-xp-highlight-unused-regexp "^[^_]")
 (racket-repl-buffer-name-function nil)
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\S ?\\S ?\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-images-system-viewer "display")
 (racket-images-system-viewer "display")
 (racket-use-repl-submit-predicate nil)
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal)))
 (racket-show-functions
  (racket-show-echo-area racket-show-pos-tip)))
(enabled-minor-modes
 (async-bytecomp-package-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (bar-cursor-mode)
 (company-box-mode)
 (company-mode)
 (company-posframe-mode)
 (counsel-mode)
 (diff-auto-refine-mode)
 (diff-hl-flydiff-mode)
 (diff-hl-mode)
 (display-line-numbers-mode)
 (doom-modeline-mode)
 (electric-indent-mode)
 (electric-pair-mode)
 (file-name-shadow-mode)
 (flycheck-mode)
 (font-lock-mode)
 (global-company-mode)
 (global-diff-hl-mode)
 (global-eldoc-mode)
 (global-flycheck-mode)
 (global-font-lock-mode)
 (global-git-commit-mode)
 (global-magit-file-mode)
 (global-origami-mode)
 (hs-minor-mode)
 (ivy-mode)
 (line-number-mode)
 (magit-auto-revert-mode)
 (mouse-wheel-mode)
 (origami-mode)
 (override-global-mode)
 (racket-xp-mode)
 (rainbow-delimiters-mode)
 (recentf-mode)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (sly-symbol-completion-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (yas-global-mode)
 (yas-minor-mode))
(disabled-minor-modes
 (abbrev-mode)
 (archive-subfile-mode)
 (auto-fill-function)
 (auto-package-update-minor-mode)
 (auto-revert-mode)
 (auto-revert-tail-mode)
 (auto-save-visited-mode)
 (blink-cursor-mode)
 (buffer-face-mode)
 (buffer-read-only)
 (cl-old-struct-compat-mode)
 (column-number-mode)
 (company-search-mode)
 (compilation-in-progress)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (defining-kbd-macro)
 (delete-selection-mode)
 (diff-hl-dir-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (display-time-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (git-commit-mode)
 (global-auto-revert-mode)
 (global-display-line-numbers-mode)
 (global-linum-mode)
 (global-page-break-lines-mode)
 (global-prettify-symbols-mode)
 (global-visual-line-mode)
 (global-whitespace-mode)
 (global-whitespace-newline-mode)
 (horizontal-scroll-bar-mode)
 (ibuffer-auto-mode)
 (ido-everywhere)
 (isearch-mode)
 (jit-lock-debug-mode)
 (linum-mode)
 (lsp-diagnostics-modeline-mode)
 (lsp-java-lens-mode)
 (lsp-lens-mode)
 (lsp-managed-mode)
 (lsp-mode)
 (lsp-origami-mode)
 (lsp-signature-mode)
 (magit-blame-mode)
 (magit-blame-read-only-mode)
 (magit-blob-mode)
 (magit-file-mode)
 (magit-wip-after-apply-mode)
 (magit-wip-after-save-local-mode)
 (magit-wip-after-save-mode)
 (magit-wip-before-change-mode)
 (magit-wip-initial-backup-mode)
 (magit-wip-mode)
 (mail-abbrevs-mode)
 (markdown-live-preview-mode)
 (menu-bar-mode)
 (mml-mode)
 (next-error-follow-minor-mode)
 (org-cdlatex-mode)
 (org-src-mode)
 (org-table-follow-field-mode)
 (orgtbl-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (page-break-lines-mode)
 (paragraph-indent-minor-mode)
 (prettify-symbols-mode)
 (projectile-mode)
 (pyvenv-mode)
 (pyvenv-tracking-mode)
 (racket-smart-open-bracket-mode)
 (server-mode)
 (sh-electric-here-document-mode)
 (shell-command-with-editor-mode)
 (size-indication-mode)
 (sly--completion-transient-mode)
 (sly-edit-value-mode)
 (sly-editing-mode)
 (sly-interactive-buttons-mode)
 (sly-macroexpansion-minor-mode)
 (sly-mode)
 (sly-popup-buffer-mode)
 (smerge-mode)
 (temp-buffer-resize-mode)
 (text-scale-mode)
 (tool-bar-mode)
 (transient-resume-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (vc-parent-buffer)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (which-function-mode)
 (whitespace-mode)
 (whitespace-newline-mode)
 (window-divider-mode)
 (with-editor-mode)
 (xref-etags-mode))
```
</details>

**Comment by greghendershott** [comment]
I'm sorry to hear that. Let's try to figure it out.

From <kbd>C-h f</kbd> `racket-xp-mode` see this section:

> When point is on a definition or use, related items are
> highlighted using racket-xp-def-face and racket-xp-use-face
> -- instead of drawing arrows as in Dr Racket. Information is
> displayed using the function(s) in the hook variable
> racket-show-functions; it is also available when hovering the
> mouse cursor.
>
> Note: If you find these point-motion features too distracting
> and/or slow, in your racket-xp-mode-hook you may disable them:
>
> #+BEGIN_SRC elisp
>   (require 'racket-xp)
>   (add-hook 'racket-xp-mode-hook
>             (lambda ()
>               (remove-hook 'pre-redisplay-functions
>                            #'racket-xp-pre-redisplay
>                            t))
> #+END_SRC
>
> The remaining features discussed below will still work.

A couple things to try:

1. Less "drastic": Experiment with the value for `racket-show-functions`. Try setting it to these values; what is the performance like, for each?

    - `'()`
    - `'(racket-show-echo-area)`
    - `'(racket-show-pos-tip)`


2. Most "drastic": Removing `racket-xp-pre-resdisplay-hook`.

**Comment by hjones2199** [comment]
Looks like it was the posframes causing the slowdown. I tried it each way and everything besides racket-show-pos-tip was just fine performance-wise. I've added (setq racket-show-functions '(racket-show-echo-area)) to my config and everything is working fine minus the pretty posframes.

Not sure why its not playing nice with posframes enabled (on two different computers running different linux distros). I tried it with emacs running in a terminal and it seemed liked performance was significantly better, which makes me think the slowdown could be related to Emacs GUI drawing the frames.

**Comment by greghendershott** [comment]
Although `pos-tip` is mostly a black box to me, the resulting OS window system tooltip has worked fine -- and fast! -- for me on Ubuntu 18.04 and on macOS. Yours is the only negative report so far -- out of an admittedly small but I think non-zero number of other users.

Even so, maybe I should _not_ default to using `pos-tip`. Instead make it opt-in.

Why I defaulted it on: Proximity helps, a lot. Having to glance back and forth between point and the echo area, is fatiguing, in my experience having dog-fooded this a lot. :(

---

I had tried using an overlay with an `after-string`, similar to how the Racket Mode debugger shows values. But in this case, having things "jump around" within the same line seemed too disorienting, so I abandoned that.

One more idea: I could try putting the annotations on complete line(s) inserted _following_ point, using e.g. `tooltip` face. I actually just now hacked together a quick experiment, and so far the experience seems nice. This would probably also be nice for displaying error messages. Because it simply uses vanilla overlays, it would be less likely (I think?) to encounter whatever issues `pos-tip` might have with OS window system tooltips.

---

TL;DR:
- Probably I should _not_ have `pos-tip` on by default.
- Possibly there might be another approach that _could_ be on by default, aside from the not-great echo area method.

**Comment by greghendershott** [comment]
For example:

![Screenshot from 2020-05-18 14-38-08](https://user-images.githubusercontent.com/265119/82248069-8b616400-9915-11ea-9e76-de8ba20091ff.png)

**Comment by greghendershott** [comment]
Or for an error message, which typically is multiple lines:

![Screenshot from 2020-05-18 14-44-55](https://user-images.githubusercontent.com/265119/82248499-3eca5880-9916-11ea-8e69-845bd4da6365.png)

Although for errors, I would probably:

- Use `error` face, unlike in that screen cap.
- Always show this, even if point not within the error span.

**Comment by greghendershott** [comment]
In fact, the more I think about it, this should probably be the _only_ approach used -- the echo area is never great, and OS tooltips cover other things and disappear eventually.

But maybe I should calm down and sleep on this. :)

**Comment by hjones2199** [comment]
`- command-execute                                                1081  53%
 - call-interactively                                            1081  53%
  - funcall-interactively                                        1081  53%
   - next-line                                                    896  44%
    - line-move                                                   896  44%
     - line-move-partial                                          823  40%
      + pos-visible-in-window-p                                    16   0%
      + posn-at-point                                               8   0%
      + window-screen-lines                                         4   0%
      + default-font-height                                         4   0%
      + scroll-up                                                   3   0%
     + line-move-visual                                            54   2%
   + counsel-M-x                                                  185   9%
- redisplay_internal (C function)                                 858  42%
 - #<compiled 0x3d124d>                                           846  41%
  - apply                                                         846  41%
   - redisplay--pre-redisplay-functions                           846  41%
    - run-hook-with-args                                          846  41%
     - racket-xp-pre-redisplay                                    846  41%
      - racket-show                                               846  41%
       - racket-show-pos-tip                                      846  41%
        - pos-tip-show                                            846  41%
         - pos-tip-show-no-propertize                             846  41%
          - x-show-tip                                            792  39%
           - face-set-after-frame-default                         792  39%
            - face-spec-recalc                                    768  38%
             + face-spec-set-2                                    411  20%
             + face-spec-reset-face                               174   8%
             + make-face-x-resource-internal                      163   8%
             + face-spec-choose                                    12   0%
            pos-tip-compute-pixel-position                         46   2%
`

**Comment by hjones2199** [comment]
I ran (profiler-start) and navigated around a .rkt file to see where the bottleneck is with some interesting results below (forgive the bad formatting). It looks like the issue could be pos-tips locking up trying to calculate fonts/faces, but its hard to tell if thats just a byproduct of slow communication between racket & emacs. I did also notice the racket pos-tips were not using the default font & color of my emacs theme.

A possible solution would be to feed the diagnostic information to flycheck-mode or flymake-mode when available. They work with different configurable frontends to display diagnostic information (like dedicated buffers, posframes, minibuffer etc), so it could respect whatever error/warning reporting method users are already using for other languages. It could fallback to minibuffer or pos-tip when they arent on the system. I'm willing to look into implementing it if its outside your interest/ballpark, but I would also be perfectly content with figuring out my pos-tip performance woes.

`
= redisplay_internal (C function)                                2025  48%
 = #<compiled 0x3d124d>                                          1994  48%
  = apply                                                        1994  48%
   = redisplay--pre-redisplay-functions                          1994  48%
    = run-hook-with-args                                         1994  48%
     = racket-xp-pre-redisplay                                   1990  48%
      = racket-show                                              1990  48%
       = racket-show-pos-tip                                     1979  47%
        = pos-tip-show                                           1979  47%
         = pos-tip-show-no-propertize                            1960  47%
          = x-show-tip                                           1804  43%
           = face-set-after-frame-default                        1801  43%
            = face-spec-recalc                                   1699  41%
             + face-spec-set-2                                    920  22%
             + face-spec-reset-face                               386   9%
             + make-face-x-resource-internal                      385   9%
             + face-spec-choose                                     4   0%
          + pos-tip-compute-pixel-position                        111   2%
            pos-tip-avoid-mouse                                     4   0%
         + pos-tip-string-width-height                              8   0%
         racket-show-echo-area                                     11   0%
 = eval                                                            27   0%
 = window--sanitize-window-sizes                                    4   0%
= command-execute                                                1912  46%
 = call-interactively                                            1908  46%
  = funcall-interactively                                        1908  46%
   = next-line                                                   1280  30%
    = line-move                                                  1280  30%
     + line-move-partial                                         1059  25%
       line-move-visual                                           129   3%

`

**Comment by greghendershott** [comment]
My `profile-report` only has lines starting with `+` or `-` -- none start with `=`.  Does that matter??

Having said that, for me, my node `x-show-tip` is empty. I don't have any of this under it (or anywhere in the report), which you have:

```
= x-show-tip 1804 43%
= face-set-after-frame-default 1801 43%
= face-spec-recalc 1699 41%
+ face-spec-set-2 920 22%
+ face-spec-reset-face 386 9%
+ make-face-x-resource-internal 385 9%
+ face-spec-choose 4 0%
+ pos-tip-compute-pixel-position 111 2%
```

Mine looks like:

```
- redisplay_internal (C function)                                1882  29%
 - #<compiled 0x43b59d>                                          1709  26%
  - apply                                                        1704  26%
   - redisplay--pre-redisplay-functions                          1695  26%
    - run-hook-with-args                                         1687  26%
     - racket-xp-pre-redisplay                                   1667  25%
      - let                                                      1667  25%
       - if                                                      1667  25%
        - let*                                                   1655  25%
         - if                                                    1651  25%
          - let                                                  1651  25%
           - racket-show                                         1641  25%
            - let                                                1637  25%
             - while                                             1637  25%
              - let                                              1633  25%
               - funcall                                         1633  25%
                - racket-show-pos-tip                            1578  24%
                 - if                                            1578  24%
                  - progn                                        1578  24%
                   - if                                          1578  24%
                    - pos-tip-show                               1573  24%
                     - pos-tip-show-no-propertize                1412  21%
                      - pos-tip-compute-pixel-position               1200  18%
                       - pos-tip-frame-top-left-coordinates                 33   0%
                        - call-process                              3   0%
                           tramp-completion-file-name-handler                  3   0%
                       - posn-at-point                              4   0%
                          eval                                      2   0%
                       - window-inside-pixel-edges                  4   0%
                          window-edges                              4   0%
                      - pos-tip-avoid-mouse                        47   0%
                         internal--before-with-selected-window                  3   0%
                      - pos-tip-compute-foreground-color                 17   0%
                         facep                                      8   0%
                      - x-show-tip                                  6   0%
                       - run-at-time                                6   0%
                          timer-activate                            3   0%
                        #<compiled 0x11a87b5>                       3   0%
                     - pos-tip-string-width-height                 13   0%
                      + #<compiled 0x11ac291>                       9   0%
                        generate-new-buffer                         4   0%
                     - pos-tip-compute-foreground-color                  4   0%
                        facep                                       4   0%
                       pos-tip-tooltip-height                       3   0%
                    + racket--non-empty-string-p                    5   0%
                + racket-show-echo-area                            55   0%
        + let                                                      12   0%
     + redisplay--update-region-highlight                          14   0%
```

I don't understand why in your case all the `face-xxx` functions are being called and are so expensive, but not in my case.

---

In any case, any mystery here makes me even more inclined to go with my proposal above, not to use `pos-tip` at all.

**Comment by greghendershott** [comment]
By the way, once the buffer is annotated (look for the check mark in the mode line), there is no further communication between Racket Mode's front and back end, while you navigate the buffer. The annotation information from the back end is in the buffer as text properties. Navigation simply uses the text properties.

Your profiler report looked like holding down <kbd>C-n</kbd> to do `next-line`. That's what I did, in a .rkt file with about 400 lines, top to bottom.

**Comment by hjones2199** [comment]
> My profile-report only has lines starting with + or - -- none start with =. Does that matter??
Mine looks like yours, I was just trying (and failing) to play nice with githubs markdown formatting (i'm more of an org-mode guy!).

I trimmed down my init.el to the basics to isolate racket-mode and the performance hit when doing `next-line` through rkt files is still very much there. Also tried the debian racket 7.2 package and racket 7.7 from nixpkgs to ensure its not a backend problem and got the same results. With that all in mind I think you're right that the culprit is pos-tip, although its odd i'm the first person to encounter (or at least report) it.

**Comment by greghendershott** [comment]
I did some more reading and learned about `x-gtk-use-system-tooltips`. For me it is `t` (by default -- nothing in my `init.el` or `custom.el` sets it).

I tried setting it to `nil`, and now I see the very bad slowdown you report.

Do you have it set to `nil`? If so, do you know why?

**Comment by hjones2199** [comment]
It's actually set to `t` on mine as well. That said, I intend to try out the new pesudo-tooltip as soon as it hits MELPA, from a glance it looks like it avoids the pos-tip pain point.

**Comment by greghendershott** [comment]
I merged to `master` -- but might be some hours before [MELPA](https://melpa.org/#/racket-mode) refreshes to show version `20200522.xxxx`.

Although I'm pretty sure that will close this issue, I'll hold off awhile to see if you have a chance to try it and have any feedback.

p.s. "Try it" = change `racket-show-functions` to e.g. `'(racket-show-pseudo-tooltip)`.

**Comment by tgbugs** [comment]
I have this issue as well using the athena gui backend where `racket-show-pos-tip` absolutely tanks performance. `(setq racket-show-functions '(racket-show-echo-area))` fixes the issue for me (yay!).

**Comment by greghendershott** [comment]
@tgbugs With latest on MELPA, you could also try `(setq racket-show-functions '(racket-show-pseudo-tooltip))`. If that doesn't perform well for you, I'd like to know!

(If you just prefer the annotations to be displayed in the echo area, that's great, too! Personally I find it annoying to keep looking down and back up again. I prefer proximity. Which is why I am being so stubborn trying to get tooltips or an equivalent working. :smile:.)

**Comment by tgbugs** [comment]
Strangely I'm running `(20200522 1724)` and I get the following error.
```
Error in post-command-hook (cursor-sensor--detect): (void-function racket-show-pseudo-tooltip)
redisplay--pre-redisplay-functions: (void-function racket-show-pseudo-tooltip) [7 times]
```

<details>
```
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
                 (20200522 1724)
                 "Racket editing, REPL, and more"
                 ((emacs
                   (25 1))
                  (faceup
                   (0 0 2))
                  (pos-tip
                   (20191127 1028)))
                 nil nil "/home/tom/.emacs.d/elpa/racket-mode-20200522.1724"
                 ((:url . "https://www.racket-mode.com/")
                  (:maintainer "Greg Hendershott")
                  (:authors
                   ("Greg Hendershott")))
                 nil))
((emacs-version "28.0.50")
 (system-type gnu/linux)
 (major-mode messages-buffer-mode)
 (racket--el-source-dir "/home/tom/git/racket-mode/")
 (racket--rkt-source-dir "/home/tom/git/racket-mode/racket/")
 (racket-program "racket")
 (racket-command-timeout 10)
 (racket-xp-after-change-refresh-delay 1)
 (racket-xp-highlight-unused-regexp "^[^_]")
 (racket-repl-buffer-name-function nil)
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\S ?\\S ?\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-images-system-viewer "display")
 (racket-images-system-viewer "display")
 (racket-use-repl-submit-predicate nil)
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal)))
 (racket-show-functions
  (racket-show-pseudo-tooltip)))
(enabled-minor-modes
 (async-bytecomp-package-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (buffer-read-only)
 (column-number-mode)
 (completion-in-region-mode)
 (csv-field-index-mode)
 (display-line-numbers-mode)
 (electric-indent-mode)
 (electric-pair-mode)
 (evil-local-mode)
 (evil-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-display-line-numbers-mode)
 (global-eldoc-mode)
 (global-font-lock-mode)
 (global-git-commit-mode)
 (global-magit-file-mode)
 (global-undo-tree-mode)
 (hexl-follow-ascii)
 (line-number-mode)
 (magit-auto-revert-mode)
 (menu-bar-mode)
 (mouse-wheel-mode)
 (override-global-mode)
 (savehist-mode)
 (semantic-minor-modes-format)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (undo-tree-mode)
 (url-handler-mode)
 (which-function-mode))
(disabled-minor-modes
 (abbrev-mode)
 (archive-subfile-mode)
 (auto-complete-mode)
 (auto-fill-function)
 (auto-insert-mode)
 (auto-revert-mode)
 (auto-revert-tail-mode)
 (auto-save-visited-mode)
 (autoarg-kp-mode)
 (autoarg-mode)
 (bibtex-completion-notes-global-mode)
 (bibtex-completion-notes-mode)
 (blink-cursor-mode)
 (buffer-face-mode)
 (bug-reference-mode)
 (bug-reference-prog-mode)
 (checkdoc-minor-mode)
 (cider--debug-mode)
 (cider-auto-test-mode)
 (cider-enlighten-mode)
 (cider-mode)
 (cider-popup-buffer-mode)
 (cl-font-lock-built-in-mode)
 (cl-old-struct-compat-mode)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (csv-align-mode)
 (cursor-intangible-mode)
 (cursor-sensor-mode)
 (cvs-minor-mode)
 (defining-kbd-macro)
 (delete-selection-mode)
 (desktop-save-mode)
 (diff-auto-refine-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (dired-isearch-filenames-mode)
 (dired-omit-mode)
 (display-battery-mode)
 (display-fill-column-indicator-mode)
 (doc-view-minor-mode)
 (doc-view-presentation-mode)
 (ede-minor-mode)
 (edebug-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (elpy-django)
 (elpy-mode)
 (emacs-lock-mode)
 (emerge-mode)
 (epa-global-mail-mode)
 (epa-mail-mode)
 (erc-button-mode)
 (erc-fill-mode)
 (erc-irccontrols-mode)
 (erc-keep-place-mode)
 (erc-list-mode)
 (erc-match-mode)
 (erc-move-to-prompt-mode)
 (erc-networks-mode)
 (erc-noncommands-mode)
 (erc-notifications-mode)
 (erc-notify-mode)
 (erc-readonly-mode)
 (erc-scrolltobottom-mode)
 (erc-services-mode)
 (erc-smiley-mode)
 (erc-stamp-mode)
 (erc-track-minor-mode)
 (erc-track-mode)
 (erc-unmorse-mode)
 (ert--current-run-stats)
 (eshell-arg-mode)
 (eshell-cmpl-mode)
 (eshell-hist-mode)
 (eshell-pred-mode)
 (eshell-proc-mode)
 (eshell-prompt-mode)
 (eshell-rebind-mode)
 (eshell-var-mode)
 (eval-sexp-fu-flash-mode)
 (evil-leader-mode)
 (evil-magit-toggle-text-minor-mode)
 (evil-org-mode)
 (evil-paredit-mode)
 (fido-mode)
 (flycheck-mode)
 (flymake-mode)
 (flyspell-mode)
 (footnote-mode)
 (gdb-enable-debug)
 (gdb-many-windows)
 (gdb-speedbar-auto-raise)
 (git-commit-mode)
 (global-auto-complete-mode)
 (global-auto-revert-mode)
 (global-display-fill-column-indicator-mode)
 (global-ede-mode)
 (global-evil-leader-mode)
 (global-flycheck-mode)
 (global-hi-lock-mode)
 (global-hl-line-mode)
 (global-hl-todo-mode)
 (global-linum-mode)
 (global-prettify-symbols-mode)
 (global-reveal-mode)
 (global-semantic-highlight-edits-mode)
 (global-semantic-highlight-func-mode)
 (global-semantic-show-parser-state-mode)
 (global-semantic-show-unmatched-syntax-mode)
 (global-semantic-stickyfunc-mode)
 (global-so-long-mode)
 (global-subword-mode)
 (global-superword-mode)
 (global-tab-line-mode)
 (global-visual-line-mode)
 (global-whitespace-mode)
 (global-whitespace-newline-mode)
 (global-wordstar-mode)
 (gnuplot-context-sensitive-mode)
 (gnus-dead-summary-mode)
 (gnus-undo-mode)
 (goto-address-mode)
 (goto-address-prog-mode)
 (gud-tooltip-mode)
 (helm--minor-mode)
 (helm--remap-mouse-mode)
 (helm-autoresize-mode)
 (helm-display-line-numbers-mode)
 (helm-migemo-mode)
 (helm-popup-tip-mode)
 (hexl-follow-ascii-mode)
 (hi-lock-mode)
 (hide-ifdef-hiding)
 (hide-ifdef-mode)
 (highlight-numbers-mode)
 (hl-line-mode)
 (hl-todo-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (html-autoview-mode)
 (ibuffer-auto-mode)
 (icomplete-mode)
 (idlwave-shell-electric-debug-mode)
 (ido-everywhere)
 (image-dired-minor-mode)
 (image-minor-mode)
 (isearch-mode)
 (ispell-minor-mode)
 (jit-lock-debug-mode)
 (js2-highlight-unused-variables-mode)
 (js2-minor-mode)
 (latex-electric-env-pair-mode)
 (linum-mode)
 (macrostep-mode)
 (magit-blame-mode)
 (magit-blame-read-only-mode)
 (magit-blob-mode)
 (magit-file-mode)
 (magit-wip-after-apply-mode)
 (magit-wip-after-save-local-mode)
 (magit-wip-after-save-mode)
 (magit-wip-before-change-mode)
 (magit-wip-initial-backup-mode)
 (magit-wip-mode)
 (mail-abbrevs-mode)
 (markdown-live-preview-mode)
 (midnight-mode)
 (minibuffer-depth-indicate-mode)
 (minibuffer-electric-default-mode)
 (mml-mode)
 (next-error-follow-minor-mode)
 (ob-ipython-mode)
 (org-capture-mode)
 (org-cdlatex-mode)
 (org-src-mode)
 (org-table-follow-field-mode)
 (orgtbl-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (paredit-mode)
 (pascal-outline-mode)
 (prettify-symbols-mode)
 (pycoverage-mode)
 (pyvenv-mode)
 (pyvenv-tracking-mode)
 (racket-smart-open-bracket-mode)
 (racket-xp-mode)
 (rainbow-delimiters-mode)
 (rainbow-delimiters-org-mode)
 (rcirc-ignore-buffer-activity-flag)
 (rcirc-low-priority-flag)
 (rcirc-multiline-minor-mode)
 (rcirc-omit-mode)
 (rcirc-track-minor-mode)
 (recentf-mode)
 (rectangle-mark-mode)
 (reftex-mode)
 (reveal-mode)
 (rng-validate-mode)
 (rst-minor-mode)
 (semantic-highlight-edits-mode)
 (semantic-highlight-func-mode)
 (semantic-mode)
 (semantic-show-parser-state-mode)
 (semantic-show-unmatched-syntax-mode)
 (semantic-stickyfunc-mode)
 (server-mode)
 (sgml-electric-tag-pair-mode)
 (sh-electric-here-document-mode)
 (shell-command-with-editor-mode)
 (size-indication-mode)
 (slime-autodoc-mode)
 (slime-edit-value-mode)
 (slime-editing-mode)
 (slime-fuzzy-target-buffer-completions-mode)
 (slime-macroexpansion-minor-mode)
 (slime-mode)
 (slime-popup-buffer-mode)
 (slime-repl-map-mode)
 (slime-repl-read-mode)
 (slime-trace-dialog-autofollow-mode)
 (slime-trace-dialog-hide-details-mode)
 (slime-trace-dialog-minor-mode)
 (smerge-mode)
 (so-long-minor-mode)
 (subword-mode)
 (superword-mode)
 (tab-bar-history-mode)
 (tab-bar-mode)
 (tab-line-mode)
 (table-fixed-width-mode)
 (table-mode-indicator)
 (tar-subfile-mode)
 (temp-buffer-resize-mode)
 (text-scale-mode)
 (tildify-mode)
 (timeclock-mode-line-display)
 (tool-bar-mode)
 (transient-resume-mode)
 (type-break-mode)
 (type-break-mode-line-message-mode)
 (type-break-query-mode)
 (undo-tree-visualizer-selection-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-dired-minor-mode)
 (use-hard-newlines)
 (vc-parent-buffer)
 (vdiff-3way-mode)
 (vdiff-mode)
 (vdiff-scroll-lock-mode)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (whitespace-mode)
 (whitespace-newline-mode)
 (window-divider-mode)
 (with-editor-mode)
 (wordstar-mode)
 (xref-etags-mode)
 (xterm-mouse-mode))
```
</details>

**Comment by greghendershott** [comment]
> I think it is the external border shifting things by a single pixel

@tgbugs I know!!! It is case where the "tooltip" doesn't extend past the end of the following line. 

Instead of the [face attribute](https://www.gnu.org/software/emacs/manual/html_node/elisp/Face-Attributes.html#Face-Attributes) `:box t` I'm deliberately using `:box (:line-width -1)`:
> :box
>     Whether or not a box should be drawn around characters, its color, the width of the box lines, and 3D appearance. Here are the possible values of the :box attribute, and what they mean:
>
>     nil
>         Don't draw a box.
>
>     t
>         Draw a box with lines of width 1, in the foreground color.
>
>     color
>         Draw a box with lines of width 1, in color color.
>
>     (:line-width width :color color :style style)
>         This way you can explicitly specify all aspects of the box.
>         The value width specifies the width of the lines to draw; it
>         defaults to 1. A negative width −n means to draw a line of
>         width n whose top and bottom parts occupy the space of the
>         underlying text, thus avoiding any increase in the character
>         height.
>
>         The value color specifies the color to draw with. The
>         default is the foreground color of the face for simple
>         boxes, and the background color of the face for 3D boxes.
>
>         The value style specifies whether to draw a 3D box. If it is
>         released-button, the box looks like a 3D button that is not
>         being pressed. If it is pressed-button, the box looks like a
>         3D button that is being pressed. If it is nil or omitted, a
>         plain 2D box is used.

But apparently that only works for the _height_ not the _width_?

I spent a highly unreasonable amount of time obsessing over how to fix that... but couldn't 100.0% and decided to set it aside.

**Comment by tgbugs** [comment]
Ugh. Been there, and have had to resist the urge to dig through emacs internals to find the actual fix. It is not an issue for me now since I have settled on using `'(racket-show-echo-area)` because I find that I like to look around the buffer near the cursor without having anything in the way, and the notifications area is a consistent place that I can always look (probably too much time spent playing RTS games with a minimap in the bottom left corner ...).

**Comment by greghendershott** [comment]
@tgbugs I did come up with some "anti-Jello", not yet merged to master: commit f4e5f75

**Comment by greghendershott** [comment]
@hjones2199 Although maybe I should wait longer to hear back from you, I think the issue is now resolved and I'm going to close it. Of course if you disagree, or have any new related problems, please feel free to re-open or open a new issue, as you prefer.
